File: CinemaReservations/CSharp/CinemaAllocations.Tests.Integration/TicketBoothShould.cs
using CinemaAllocations.Domain;
using NFluent;
using Xunit;

namespace CinemaAllocations.Tests.Integration
{
    public class TicketBoothShould
    {
        [Fact]
        public void Reserve_one_seat_when_available()
        {
            const int partyRequested = 1;

            using (var repository = Given.The.FordTheater)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.FordTheaterId, partyRequested));

                Check.That(seatsAllocated.ReservedSeats).HasSize(1);
                Check.That(seatsAllocated.ReservedSeats[0].ToString()).IsEqualTo("A3");
            }
        }

        [Fact]
        public void Reserve_multiple_seats_when_available()
        {
            const int partyRequested = 3;

            using (var repository = Given.The.DockStreet)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.DockStreetId, partyRequested));

                Check.That(seatsAllocated.ReservedSeats).HasSize(3);
                Check.That(seatsAllocated.SeatNames()).ContainsExactly("A6", "A7", "A8");
            }
        }

        [Fact]
        public void Return_SeatsNotAvailable_when_all_seats_are_unavailable()
        {
            const int partyRequested = 1;

            using (var repository = Given.The.MadisonTheather)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.MadisonTheatherId, partyRequested));

                Check.That(seatsAllocated).IsInstanceOf<NoPossibleAllocationsFound>();
            }
        }

        [Fact]
        public void Return_TooManyTicketsRequested_when_9_tickets_are_requested()
        {
            const int partyRequested = 9;

            using (var repository = Given.The.MadisonTheather)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.MadisonTheatherId, partyRequested));

                Check.That(seatsAllocated).IsInstanceOf<TooManyTicketsRequested>();
            }
        }

        [Fact]
        public void Reserve_three_adjacent_seats_when_available()
        {
            const int partyRequested = 3;

            using (var repository = Given.The.O3Auditorium)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.O3AuditoriumId, partyRequested));

                Check.That(seatsAllocated.ReservedSeats).HasSize(3);
                Check.That(seatsAllocated.SeatNames()).ContainsExactly("A8", "A9", "A10");
            }
        }

        [Fact]
        public void Return_NoPossibleAdjacentSeatsFound_when_4_tickets_are_requested()
        {
            const int partyRequested = 4;

            using(var repository = Given.The.O3Auditorium)
            {
                var ticketBooth = new TicketBooth(repository);

                var seatsAllocated =
                    ticketBooth.AllocateSeats(new AllocateSeats(Given.The.O3AuditoriumId, partyRequested));

                Check.That(seatsAllocated).IsInstanceOf<NoPossibleAdjacentSeatsFound>();
            }
        }
    }
}

File: CinemaReservations/Java/CinemaAllocationsInfra/src/test/java/com/sdd/cinemaallocationsinfra/repository/TicketboothShould.java
package com.sdd.cinemaallocationsinfra.repository;

import com.sdd.cinemaallocations.*;
import com.sdd.cinemaallocationsinfra.helpers.Given;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.context.junit4.SpringRunner;

import javax.persistence.EntityManager;
import java.io.IOException;

import static com.google.common.truth.Truth.assertThat;

@RunWith(SpringRunner.class)
@DataJpaTest
public class TicketboothShould {

    @Autowired
    private EntityManager entityManager;

    @Autowired
    private SpringMovieScreeningRepository springMovieScreeningRepository;

    @Test
    public void reserve_one_seat_when_available() throws IOException {

        TicketBooth ticketBooth = Given.The.fordTheaterTicketBooth(springMovieScreeningRepository);
        int partyRequested = 1;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.ford_theater_id, partyRequested));

        assertThat(seatsAllocated.reservedSeats()).hasSize(1);
        assertThat(seatsAllocated.reservedSeats().get(0).toString()).isEqualTo("A3");

    }

    @Test
    public void Reserve_multiple_seats_when_available() throws IOException {
        TicketBooth ticketBooth = Given.The.dockStreetTicketBooth(springMovieScreeningRepository);

        int partyRequested = 3;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.dock_street_id, partyRequested));

        assertThat(seatsAllocated.reservedSeats()).hasSize(3);
        assertThat((seatsAllocated.seatNames())).containsExactly("A6", "A7", "A8");
    }

    @Test
    public void return_SeatsNotAvailable_when_all_seats_are_unavailable() throws IOException {
        TicketBooth ticketBooth = Given.The.madisonTheaterTicketBooth(springMovieScreeningRepository);

        int partyRequested = 1;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.madison_theater_id, partyRequested));

        assertThat(seatsAllocated).isInstanceOf(NoPossibleAllocationsFound.class);
    }

    @Test
    public void return_TooManyTicketsRequested_when_9_tickets_are_requested() throws IOException {
        TicketBooth ticketBooth = Given.The.madisonTheaterTicketBooth(springMovieScreeningRepository);

        int partyRequested = 9;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.madison_theater_id, partyRequested));

        assertThat(seatsAllocated).isInstanceOf(TooManyTicketsRequested.class);
    }

    @Test
    public void reserve_three_adjacent_seats_when_available() throws IOException {
        TicketBooth ticketBooth = Given.The.o3AuditoriumTicketBooth(springMovieScreeningRepository);

        int partyRequested = 3;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.os_auditorium_id, partyRequested));

        assertThat(seatsAllocated.reservedSeats()).hasSize(3);
        assertThat((seatsAllocated.seatNames())).containsExactly("A8", "A9", "A10");
    }

    @Test
    public void return_NoPossibleAdjacentSeatsFound_when_4_tickets_are_requested() throws IOException {
        TicketBooth ticketBooth = Given.The.o3AuditoriumTicketBooth(springMovieScreeningRepository);

        int partyRequested = 4;

        SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(new AllocateSeats(Given.The.os_auditorium_id, partyRequested));

        assertThat(seatsAllocated).isInstanceOf(NoPossibleAdjacentSeatsFound.class);
    }
}
